File: CPP/GenericTypes.cpp

#include <iostream>
#include <string>
using namespace std;

// generic type class
template<class T>
class Node {
private:
    T data;
public:
    Node(const T& data) {
        this->data = data;
    }
    T getData() const {
        return this->data;
    }
    void setData(const T& data) {
        this->data = data;
    }
};

// generic type function
template<typename T>
T findSmaller(const T& a, const T& b) {
    if (a <= b) {
        return a;
    }
    return b;
}

int main(int argc, char const *argv[])
{
    // using user-defined generic datatype
    Node<int> intNode(3);
    cout << intNode.getData() << endl;
    Node<string> strNode("Hello");
    cout << strNode.getData() << endl;

    // generic type function
    int a = 3;
    int b = 4;
    cout << findSmaller(a, b) << endl;
    // generic type function given datatypes that do not have "<=" will not compile
    // Node<int> intNode2(4);
    // findSmaller(intNode, intNode2);
    
    return 0;
}

File: Java/GenericTypes.java
package Java;
// generic type class: Java/Node.java

public class GenericTypes {

    public static void main(String[] args) {
        // user-defined generic class with different datatypes
        Node<Integer> intNode = new Node<>(3);
        System.out.println(intNode.data());
        Node<Boolean> boolNode = new Node<>(true);
        System.out.println(boolNode.data());
        // generic type function
        double a = 2.4;
        double b = 2.39;
        System.out.println(findSmaller(a, b));
        String strA = "abc";
        String strB = "cde";
        System.out.println(findSmaller(strA, strB));
    }

    // This will not compile, because type T does not have "<=" operator
    // public static <T> T findSmaller(T a, T b) {
    //     if (a <= b) {
    //         return a;
    //     } 
    //     return b;
    // } 
    
    // This compiles, using Comparable's compareTo method
    public static <T extends Comparable<T>> T findSmaller(T a, T b) {
        if (a.compareTo(b) <= 0) {
            return a;
        } 
        return b;
    }    
    
}